### PR TITLE
Fix issue #141 

### PR DESCRIPTION
**Pull request by saullocarvalho**
Fix issue #141 .

I've converted the iterated AST processing algorithm to a recursive one, and I've modified `shallow_repr` to consider infix operation precedence level.

### CODE REVIEW

**Comment by zardus** [comment]
@rhelmot, do you want to take a look to see if this is what you envisioned? I'm ready to merge this otherwise as a solid start to this concept!

**Comment by zardus** [comment]
(whenever you have a moment, ofc)

File: tests/test_ast.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import nose.tools
+
+import claripy
+
+
+def test_lite_repr():
+    one = claripy.BVV(1, 8)
+    two = claripy.BVV(2, 8)
+    a = claripy.BVS('a', 8)
+    b = claripy.BVS('b', 8)
```
Comment (rhelmot) [inline tests/test_ast.py]: you can avoid the regex assert (please avoid the regex assert) by either templating the target string or using `explicit_name=True` in the BVS constructor.

File: claripy/ast/base.py [hunk 2]
```diff
@@ -197,10 +198,11 @@ def _calc_hash(op, args, keywords):
...
         Calculates the hash of an AST, given the operation, args, and kwargs.
 
-        :param op:          The operation.
-        :param args:        The arguments to the operation.
-        :param keywords:    A dict including the 'symbolic', 'variables', and 'length' items.
-        :returns:           a hash.
+        :param op:                  The operation.
+        :param args:                The arguments to the operation.
+        :param keywords:            A dict including the 'symbolic', 'variables', and 'length' items.
+
```
Comment (rhelmot) [inline claripy/ast/base.py]: I would prefer not to have a blank line in between the parameters and the return value. rst will treat the list as two separate lists when the newline is inserted.

File: tests/test_ast.py (hunk 1 above)
Comment (saullocarvalho) [reply tests/test_ast.py]: I didn't know about `explicit_name=True` use. Thanks for the information!
I've just made the change.

File: claripy/ast/base.py (hunk 2 above)
Comment (saullocarvalho) [reply claripy/ast/base.py]: Blank lines removed!
I am going to fix it on angr-docs. The blank line between parameters and the return value seems expected.

**Comment by saullocarvalho** [comment]
I've added tests to check left-to-right associativity.

File: tests/test_ast.py [hunk 3]
```diff
@@ -6,19 +6,31 @@
...
 
 
+def test_associativity():
+    x = claripy.BVS('x', 8, explicit_name=True)
+    y = claripy.BVS('y', 8, explicit_name=True)
+    z = claripy.BVS('z', 8, explicit_name=True)
+    w = claripy.BVS('w', 8, explicit_name=True)
+
+    nose.tools.assert_equal((x - y - z - w).shallow_repr(),
+                            (x - (y - (z - w))).shallow_repr())
```
Comment (rhelmot) [inline tests/test_ast.py]: this test is not correct! `x - y - z - w` is the same as `+x + -y + -z + -w`, while `x - (y - (z - w))` is the same as `+x + -y + +z + -w`

File: tests/test_ast.py [hunk 4]
```diff
@@ -6,19 +6,31 @@
...
-    nose.tools.assert_regex((a + one * b + two).shallow_repr(),
-                            r'\<BV8 __add__\(a_\d+_8, 1 \* b_\d+_8, 2\)\>')
-    nose.tools.assert_regex(((a + one) * (b + two)).shallow_repr(),
-                            r'\<BV8 \(a_\d+_8 \+ 1\) \* \(b_\d+_8 \+ 2\)\>')
-    nose.tools.assert_regex((a * one + b * two).shallow_repr(),
-                            r'\<BV8 a_\d+_8 \* 1 \+ b_\d+_8 \* 2\>')
-    nose.tools.assert_regex(((one + a)*(two + b)+(two + a)*(one + b)).shallow_repr(),
-                            r'\<BV8 \(1 \+ a_\d+_8\) \* \(2 \+ b_\d+_8\) \+ \(2 \+ a_\d+_8\) \* \(1 \+ b_\d+_8\)\>')
+    nose.tools.assert_equal((a + one * b + two).shallow_repr(),
+                            '<BV8 __add__(a, 1 * b, 2)>')
```
Comment (rhelmot) [inline tests/test_ast.py]: can we make it so that this renders as `a + 1 * b + 2`?

File: tests/test_ast.py (hunk 3 above)
Comment (saullocarvalho) [reply tests/test_ast.py]: Oops! You're right. I've missed that detail. I am gonna fix it.

File: tests/test_ast.py (hunk 4 above)
Comment (saullocarvalho) [reply tests/test_ast.py]: Yes. We can.

Comment (saullocarvalho) [reply tests/test_ast.py]: The infix operator with more than 2 args is full represented. Even when using `details == Base.LITE_REPR`. I've changed this.

File: tests/test_ast.py (hunk 3 above)
Comment (saullocarvalho) [reply tests/test_ast.py]: I've misunderstood your previous comment. Thanks for pointing the mistake. I've fixed the associativity tests and the code to handle it correctly.

**Comment by rhelmot** [comment]
absolutely beautiful. thank you!

**Pull request merged**